indexing-service/src/test/java/org/apache/druid/indexing/overlord/MSQControllerTaskLimitTest.java

```diff
@@ -0,0 +1,193 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.overlord;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.druid.indexing.common.actions.TaskAction;
+import org.apache.druid.indexing.common.actions.TaskActionClient;
+import org.apache.druid.indexing.common.actions.TaskActionClientFactory;
+import org.apache.druid.indexing.common.config.TaskStorageConfig;
+import org.apache.druid.indexing.common.task.NoopTask;
+import org.apache.druid.indexing.overlord.config.DefaultTaskConfig;
+import org.apache.druid.indexing.overlord.config.TaskLockConfig;
+import org.apache.druid.indexing.overlord.config.TaskQueueConfig;
+import org.apache.druid.java.util.common.io.Closer;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.metadata.IndexerSQLMetadataStorageCoordinator;
+import org.apache.druid.metadata.TestDerbyConnector;
+import org.apache.druid.segment.TestHelper;
+import org.apache.druid.server.metrics.NoopServiceEmitter;
+import org.joda.time.Period;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class MSQControllerTaskLimitTest
+{
+  private static final String DATASOURCE = "ds";
+
+
+  @Rule
+  public final TestDerbyConnector.DerbyConnectorRule derbyConnectorRule = new TestDerbyConnector.DerbyConnectorRule();
+
+  private TaskQueue taskQueue;
+  private TaskStorage taskStorage;
+  private SimpleTaskRunner taskRunner;
+  private int numTasks = 3;
+  private Closer closer;
+
+  @Before
+  public void setUp()
+  {
+    EmittingLogger.registerEmitter(new NoopServiceEmitter());
+
+    closer = Closer.create();
+
+    // Be as realistic as possible; use actual classes for storage rather than mocks.
+    taskStorage = new HeapMemoryTaskStorage(new TaskStorageConfig(Period.hours(1)));
+    taskRunner = new SimpleTaskRunner(3);
+    closer.register(taskRunner::stop);
+  }
+
+  @After
+  public void tearDown() throws Exception
+  {
+    closer.close();
+  }
+
+  @Test
+  public void testMaxControllerTasksLimit() throws InterruptedException
+  {
+
+    startTaskQueue(new TaskQueueConfig(null, Period.millis(1), null, null, null, null, 2));
+
+    Assert.assertEquals("no tasks should be running", 0, taskRunner.getKnownTasks().size());
+    Assert.assertEquals("no tasks should be known", 0, taskQueue.getTasks().size());
+    Assert.assertEquals("no tasks should be running", 0, taskQueue.getRunningTaskCount().size());
+
+    // Add all tasks.
+    for (int i = 0; i < numTasks; i++) {
+      final NoopTask testTask = createMsqTestTask(1000L);
+      taskQueue.add(testTask);
+    }
+
+    Thread.sleep(3000);
```

Comment:
Is it possible to mark tasks complete manually. `Thread.sleep` would not be great idea here, I think

Reply:
This is more less how it's done in TaskQueueScaleTest, I'm not sure how to make it manageable as we are testing the way how taskQueue is calling runner.run()

Reply:
Yes, I checked other places as well, and `Thread.sleep` is used there too. So I believe we’re good. For example:

```
private void waitUntil(KinesisIndexTask task, Predicate<KinesisIndexTask> predicate)
      throws InterruptedException
  {
    while (!predicate.test(task)) {
      Thread.sleep(10);
    }
  }
  ```